rllib/core/optim/tests/test_rl_optimizer_tf.py

```diff
@@ -0,0 +1,477 @@
+    def _do_update_fn(self, batch: SampleBatch) -> Mapping[str, Any]:
+        with tf.GradientTape() as tape:
+            # fwd_out = self._module.forward_train(batch)
+            fwd_out = self._module(batch)
+            loss = self._rl_optimizer.compute_loss(batch, fwd_out)
+        gradients = self.compute_gradients(loss, tape)
+        self.apply_gradients(gradients)
+        return {"loss": loss, "fwd_out": fwd_out, "post_processed_gradients": gradients}
+
+    def update(self, batch: SampleBatch) -> Mapping[str, Any]:
+        """Perform an update on this Trainer.
+
+        Args:
+            batch: A batch of data.
+
+        Returns:
+            A dictionary of results.
+        """
+        for key, value in batch.items():
+            batch[key] = tf.convert_to_tensor(value, dtype=tf.float32)
+        # with tf.GradientTape() as tape:
+        #     fwd_out = self._module.forward_train(batch)
+        #     loss = self._rl_optimizer.compute_loss(batch, fwd_out)
+        # gradients = self.compute_gradients(loss, tape)
+        # post_processed_gradients = self.on_after_compute_gradients(gradients)
+        # self.apply_gradients(post_processed_gradients)
+        # tf.config.run_functions_eagerly(True)
+        infos = self._do_update_fn(batch)
+        # assert False
+        loss = infos["loss"]
+        fwd_out = infos["fwd_out"]
+        post_processed_gradients = infos["post_processed_gradients"]
+        results = self.compile_results(batch, fwd_out, loss, post_processed_gradients)
+        return results
+
+    def compute_gradients(
+        self, loss: Union[TensorType, Mapping[str, Any]], tape: tf.GradientTape
+    ) -> Mapping[str, Any]:
+        """Perform an update on self._module
+
+            For example compute and apply gradients to self._module if
+                necessary.
+
+        Args:
+            loss: variable(s) used for optimizing self._module.
+
+        Returns:
+            A dictionary of extra information and statistics.
+        """
+        grads = tape.gradient(loss["total_loss"], self._module.trainable_variables())
+        return grads
+
+    def apply_gradients(self, gradients: Mapping[str, Any]) -> None:
+        """Perform an update on self._module"""
+        for key, optimizer in self._rl_optimizer.get_optimizers().items():
+            optimizer.apply_gradients(
+                zip(gradients[key], self._module.trainable_variables()[key])
+            )
+
+    def set_state(self, state: Mapping[str, Any]) -> None:
+        """Set the state of the trainer."""
+        self._rl_optimizer.set_state(state.get("optimizer_state", {}))
+        self._module.set_state(state.get("module_state", {}))
+
+    def get_state(self) -> Mapping[str, Any]:
+        """Get the state of the trainer."""
+        return {
+            "module_state": self._module.get_state(),
+            "optimizer_state": self._rl_optimizer.get_state(),
+        }
+
+
+class TestRLOptimizerTF(unittest.TestCase):
```

Comment:
Looking at these tests, I feel like we should just create one `test_rl_optimizer.py` and then have that one sweep through a two element list of triplets of (trainers, rl_modules, and rl_optimizers) for torch/tf. The rest of what we do is the same.

Reply:
I think we should do this once we merge the RLTrainer API.

Also the set_state/get_state tests are different

Reply:
sure. We can do it in a separate PR.